File: binding/include/multimedia/libqsoundeffect.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsoundeffect.html
pub const qsoundeffect = struct {

    /// New constructs a new QSoundEffect object.
    pub fn New() QtC.QSoundEffect {
        return qtc.QSoundEffect_new();
    }


    /// New2 constructs a new QSoundEffect object.
    pub fn New2(audioDevice: ?*anyopaque) QtC.QSoundEffect {
        return qtc.QSoundEffect_new2(@ptrCast(audioDevice));
    }


    /// New3 constructs a new QSoundEffect object.
    pub fn New3(parent: ?*anyopaque) QtC.QSoundEffect {
        return qtc.QSoundEffect_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QSoundEffect object.
    pub fn New4(audioDevice: ?*anyopaque, parent: ?*anyopaque) QtC.QSoundEffect {
        return qtc.QSoundEffect_new4(@ptrCast(audioDevice), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSoundEffect_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsoundeffect.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SupportedMimeTypes() []const u8 {
        return qtc.QSoundEffect_SupportedMimeTypes();
    }

    pub fn Source(self: ?*anyopaque, ) QtC.QUrl {
        return qtc.QSoundEffect_Source(@ptrCast(self));
    }

    pub fn SetSource(self: ?*anyopaque, url: ?*anyopaque) void {
        qtc.QSoundEffect_SetSource(@ptrCast(self), @ptrCast(url));
    }

    pub fn LoopCount(self: ?*anyopaque, ) i32 {
        return qtc.QSoundEffect_LoopCount(@ptrCast(self));
    }

    pub fn LoopsRemaining(self: ?*anyopaque, ) i32 {
        return qtc.QSoundEffect_LoopsRemaining(@ptrCast(self));
    }

    pub fn SetLoopCount(self: ?*anyopaque, loopCount: i32) void {
        qtc.QSoundEffect_SetLoopCount(@ptrCast(self), loopCount);
    }

    pub fn AudioDevice(self: ?*anyopaque, ) QtC.QAudioDevice {
        return qtc.QSoundEffect_AudioDevice(@ptrCast(self));
    }

    pub fn SetAudioDevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QSoundEffect_SetAudioDevice(@ptrCast(self), @ptrCast(device));
    }

    pub fn Volume(self: ?*anyopaque, ) f32 {
        return qtc.QSoundEffect_Volume(@ptrCast(self));
    }

    pub fn SetVolume(self: ?*anyopaque, volume: f32) void {
        qtc.QSoundEffect_SetVolume(@ptrCast(self), volume);
    }

    pub fn IsMuted(self: ?*anyopaque, ) bool {
        return qtc.QSoundEffect_IsMuted(@ptrCast(self));
    }

    pub fn SetMuted(self: ?*anyopaque, muted: bool) void {
        qtc.QSoundEffect_SetMuted(@ptrCast(self), muted);
    }

    pub fn IsLoaded(self: ?*anyopaque, ) bool {
        return qtc.QSoundEffect_IsLoaded(@ptrCast(self));
    }

    pub fn IsPlaying(self: ?*anyopaque, ) bool {
        return qtc.QSoundEffect_IsPlaying(@ptrCast(self));
    }

    pub fn Status(self: ?*anyopaque, ) i32 {
        return qtc.QSoundEffect_Status(@ptrCast(self));
    }

    pub fn SourceChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_SourceChanged(@ptrCast(self));
    }

    pub fn OnSourceChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_SourceChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LoopCountChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_LoopCountChanged(@ptrCast(self));
    }

    pub fn OnLoopCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_LoopCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LoopsRemainingChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_LoopsRemainingChanged(@ptrCast(self));
    }

    pub fn OnLoopsRemainingChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_LoopsRemainingChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn VolumeChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_VolumeChanged(@ptrCast(self));
    }

    pub fn OnVolumeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_VolumeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MutedChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_MutedChanged(@ptrCast(self));
    }

    pub fn OnMutedChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_MutedChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LoadedChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_LoadedChanged(@ptrCast(self));
    }

    pub fn OnLoadedChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_LoadedChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn PlayingChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_PlayingChanged(@ptrCast(self));
    }

    pub fn OnPlayingChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_PlayingChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn StatusChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_StatusChanged(@ptrCast(self));
    }

    pub fn OnStatusChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_StatusChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn AudioDeviceChanged(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_AudioDeviceChanged(@ptrCast(self));
    }

    pub fn OnAudioDeviceChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QSoundEffect_Connect_AudioDeviceChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Play(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_Play(@ptrCast(self));
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QSoundEffect_Stop(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSoundEffect_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsoundeffect.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSoundEffect_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsoundeffect.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSoundEffect_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/soundeffect.html#types
pub const enums = struct {
    pub const Loop = enum {
        pub const Infinite: i32 = -2;
    };

    pub const Status = enum {
        pub const Null: i32 = 0;
        pub const Loading: i32 = 1;
        pub const Ready: i32 = 2;
        pub const Error: i32 = 3;
    };

};

File: binding/src/multimedia/libqsoundeffect.cpp
#include <QAudioDevice>
#include <QObject>
#include <QSoundEffect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <qsoundeffect.h>
#include "libqsoundeffect.h"
#include "libqsoundeffect.hxx"

QSoundEffect* QSoundEffect_new() {
	 return new QSoundEffect();
}

QSoundEffect* QSoundEffect_new2(const QAudioDevice* audioDevice) {
	 return new QSoundEffect(*audioDevice);
}

QSoundEffect* QSoundEffect_new3(QObject* parent) {
	 return new QSoundEffect(parent);
}

QSoundEffect* QSoundEffect_new4(const QAudioDevice* audioDevice, QObject* parent) {
	 return new QSoundEffect(*audioDevice, parent);
}

libqt_string QSoundEffect_Tr(const char* s) {
	QString _ret = QSoundEffect::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QSoundEffect_SupportedMimeTypes() {
	return QSoundEffect::supportedMimeTypes();
}

QUrl* QSoundEffect_Source(const QSoundEffect* self) {
	return new QUrl(self->source());
}

void QSoundEffect_SetSource(QSoundEffect* self, const QUrl* url) {
	self->setSource(*url);
}

int QSoundEffect_LoopCount(const QSoundEffect* self) {
	return self->loopCount();
}

int QSoundEffect_LoopsRemaining(const QSoundEffect* self) {
	return self->loopsRemaining();
}

void QSoundEffect_SetLoopCount(QSoundEffect* self, int loopCount) {
	self->setLoopCount(loopCount);
}

QAudioDevice* QSoundEffect_AudioDevice(QSoundEffect* self) {
	return new QAudioDevice(self->audioDevice());
}

void QSoundEffect_SetAudioDevice(QSoundEffect* self, const QAudioDevice* device) {
	self->setAudioDevice(*device);
}

float QSoundEffect_Volume(const QSoundEffect* self) {
	return self->volume();
}

void QSoundEffect_SetVolume(QSoundEffect* self, float volume) {
	self->setVolume(volume);
}

bool QSoundEffect_IsMuted(const QSoundEffect* self) {
	return self->isMuted();
}

void QSoundEffect_SetMuted(QSoundEffect* self, bool muted) {
	self->setMuted(muted);
}

bool QSoundEffect_IsLoaded(const QSoundEffect* self) {
	return self->isLoaded();
}

bool QSoundEffect_IsPlaying(const QSoundEffect* self) {
	return self->isPlaying();
}

int QSoundEffect_Status(const QSoundEffect* self) {
	return self->status();
}

void QSoundEffect_SourceChanged(QSoundEffect* self) {
	self->sourceChanged();
}

void QSoundEffect_Connect_SourceChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::sourceChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_LoopCountChanged(QSoundEffect* self) {
	self->loopCountChanged();
}

void QSoundEffect_Connect_LoopCountChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::loopCountChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_LoopsRemainingChanged(QSoundEffect* self) {
	self->loopsRemainingChanged();
}

void QSoundEffect_Connect_LoopsRemainingChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::loopsRemainingChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_VolumeChanged(QSoundEffect* self) {
	self->volumeChanged();
}

void QSoundEffect_Connect_VolumeChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::volumeChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_MutedChanged(QSoundEffect* self) {
	self->mutedChanged();
}

void QSoundEffect_Connect_MutedChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::mutedChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_LoadedChanged(QSoundEffect* self) {
	self->loadedChanged();
}

void QSoundEffect_Connect_LoadedChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::loadedChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_PlayingChanged(QSoundEffect* self) {
	self->playingChanged();
}

void QSoundEffect_Connect_PlayingChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::playingChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_StatusChanged(QSoundEffect* self) {
	self->statusChanged();
}

void QSoundEffect_Connect_StatusChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::statusChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_AudioDeviceChanged(QSoundEffect* self) {
	self->audioDeviceChanged();
}

void QSoundEffect_Connect_AudioDeviceChanged(QSoundEffect* self, intptr_t slot) {
    void (*slotFunc)(QSoundEffect*) = reinterpret_cast<void (*)(QSoundEffect*)>(slot);
    QSoundEffect::connect(self, &QSoundEffect::audioDeviceChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QSoundEffect_Play(QSoundEffect* self) {
	self->play();
}

void QSoundEffect_Stop(QSoundEffect* self) {
	self->stop();
}

libqt_string QSoundEffect_Tr2(const char* s, const char* c) {
	QString _ret = QSoundEffect::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSoundEffect_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSoundEffect::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSoundEffect_Delete(QSoundEffect* self) {
    delete self;
}

